src/GuzzleMiddlewareFactory.php

```diff
@@ -0,0 +1,141 @@
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Log\LoggerInterface;
+use Sentry\SentrySdk;
+
+class GuzzleMiddlewareFactory
+{
+    private const MAX_LOG_BODY_IN_CHARS = 200;
+
+    private int $maxBody;
+
+    public function __construct(private LoggerInterface $logger, YiiSentryConfig $conf)
+    {
+        $this->maxBody = $conf->getMaxGuzzleBodyTrace() ?? self::MAX_LOG_BODY_IN_CHARS;
+    }
+
+    public function factory(callable $handler): callable
+    {
+        return function (RequestInterface $srcRequest, array $options) use ($handler) {
+            $traceHeader = SentrySdk::getCurrentHub()->getSpan()?->toTraceparent();
+            if ($traceHeader) {
+                $request = $srcRequest->withAddedHeader('sentry-trace', $traceHeader);
+            } else {
+                $request = clone $srcRequest;
+            }
+
+            $startTime = microtime(true);
+            $path = $request->getMethod() . ':' . $request->getUri()->__toString();
+            $requestBody = $request->getBody()->isReadable()
+                ? $request->getBody()->getContents()
+                : '[not readable]';
+            if (mb_strlen($requestBody) > $this->maxBody) {
+                $requestContentBody = mb_strimwidth($requestBody, 0, $this->maxBody, "...");
+            } else {
+                $requestContentBody = $requestBody;
+            }
+            /** @var PromiseInterface $response */
+            $response = $handler($request, $options);
+
+            $requestMethod = $request->getMethod();
+            $requestHeaders = $request->getHeaders();
+
+            return $response->then(
+                function (ResponseInterface $promiseResponse) use (
+                    $startTime,
+                    $requestContentBody,
+                    $path,
+                    $requestHeaders,
+                    $requestMethod,
+                ) {
+                    $responseContentBody = $this->getResponseContentBody($promiseResponse);
+                    $logContext = [
+                    'time' => $startTime,
+                    'elapsed' => microtime(true) - $startTime,
+                    'category' => 'guzzle.request',
+                    'method' => $requestMethod,
+                    'request_headers' => $requestHeaders,
+                    'response_headers' => $promiseResponse->getHeaders(),
+                    'request_body' => $requestContentBody,
+                    'response_body' => $responseContentBody,
+                    ];
+                    $this->logger->info($path, $logContext);
+
+                    return $promiseResponse;
+                }, function (Exception $e) use (
+                    $startTime,
+                    $requestContentBody,
+                    $path,
+                ) {
+                    if ($e instanceof RequestException) {
+                        $responseContentBody = $this->getResponseContentBody($e->getResponse());
```

Comment:
Common key / value pairs. such as `time`, `elapsed`, etc. can be combined into one array and then merged with unique ones.